### PR TITLE
fix(iam): no trust policy added for cross-account roles

### LINKED ISSUE #5740: S3 Bucket grant* stopped working in 1.20.0
I've upgraded cdk (and aligned their modules as well) from 1.16.3(?) to 1.20.0 and I noticed that bucket policies that I created with the following stopped working
```javascript
myBucket.grantReadWrite(myRole, `*/${templatesPublic}/*`);
```

### Reproduction Steps

What I saw with `cdk diff` 1.20.0:
image
image

If I renamed the bucket, I'd get:
image
_So, it suggests that it's broken for new and existing buckets as well._

It's not just the diff that was wrong; when I did `cdk deploy`, the bucket policy was actually gone.

Downgrading cdk (and the node modules) to 1.19.0:
I saw iam policies are back to what I had before:
image
image

### Environment

  - **CLI Version      :*1.20.0*
  - **Framework Version:**
  - **OS               :*Linux*
  - **Language         :*JavaScript*

I didn't spot any breaking change notice in the release notes at:
https://github.com/aws/aws-cdk/releases/tag/v1.20.0

--- 

This is :bug: Bug Report

### PR DESCRIPTION
**Pull request by rix0rrr**
Trust policies should be added to resources when trying to grant
permissions to a principal in a different account.  Typically this is a
Service Principal, but it could also be an imported Role from a
different account (and the same should not be done for a Role in the
same account).

Our previous API (`addToPolicy` returning `true|false`) was designed for
in-account Roles and cross-account ServicePrincipals, but was not rich
enough to make the distinction that Roles could be cross-account.

Add a function to `IPrincipal` to ask it whether it is in the same or a
different account from a given scope, and have `Grant` respect that.

Also in this commit: a bugfix for `ImmutableRole` to make it work
correctly with `Grant`

Also in this commit: `parseArn()` used to behave the same for
`parseArn(TOKEN)` and `parseArn("arn:{TOKEN}:svc:1234:...")`.
Make it (opportunistically) do the right thing in the latter case,
parsing out the concrete components and the TOKEN components
separately. This makes it possible to parse out the concrete
account number from an ARN that contains tokenized parts we don't
care about.

Fixes #5740, #2807.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by rix0rrr** [comment]
This change might be contentious, in the addition of `sameAccount()?` and the behavior change of `parseArn()` (both of which I feel are defensible but also a reasonable person might disagree with the choices made).

Would appreciate some advice here.

Also see the set of test cases for coverage.

File: packages/@aws-cdk/aws-iam/test/policy-document.test.ts [hunk 1]
```diff
@@ -341,7 +341,8 @@ describe('IAM polocy document', () => {
       get grantPrincipal() { return this; },
       assumeRoleAction: 'sts:AssumeRole',
       policyFragment: new PrincipalPolicyFragment({ AWS: ['foo', 'bar'] }),
-      addToPolicy() { return false; }
+      addToPolicy() { return false; },
+      sameAccount() { return false; }
```
Comment (skinny85) [inline packages/@aws-cdk/aws-iam/test/policy-document.test.ts]: Trailing comma, to avoid the diff problem in the future? :)

**Comment by rix0rrr** [comment]
> I've actually changed my mind on this, and I want to explore an alternative solution. I think we should develop something here that's usable across all construct libraries, not specific to IAM.

I'm always in favor of better solutions for anything. As long as we can solve the same issue as mentioned here, go ahead.

> We have a related issue about this around the necessity of having access to `account` and `region` on a resource level, instead of just stack level, but I can't find it :/.

I guess because of the nature of our deployment mechanism:

* Environment of "owned" resources == environment of containing Stack
* Environment of "imported" resources == ¯\_(ツ)_/¯ lol can be anything look at the ARN

That seems to be the crux of it, and you're right it holds for all imported resources, not just roles.

> @rix0rrr are you OK not merging this for the time being, and giving me some time to work on it?

As long as you incorporate the tests of this PR into whatever you come up with, sure!

File: packages/@aws-cdk/core/lib/cfn-pseudo.ts [hunk 2]
```diff
@@ -77,5 +77,5 @@ export class ScopedAws {
 }
 
 function pseudoString(name: string): string {
-  return Token.asString({ Ref: name }, { displayHint: name });
+  return Token.asString({ Ref: name }, { displayHint: name.replace('::', '.') });
```
Comment (eladb) [inline packages/@aws-cdk/core/lib/cfn-pseudo.ts:80]: why is this needed?

File: packages/@aws-cdk/aws-iam/lib/private/accounts.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+  return !Token.isUnresolved(account1) && !Token.isUnresolved(account2) && account1 !== account2;
+}
+
+/**
+ * Return whether two accounts are the same account
+ *
+ * Returns undefined if one is agnostic and the other one isn't.
+ */
+export function sameAccount(account1: string | undefined, account2: string | undefined): boolean | undefined {
+  // Both agnostic in 99% of cases means they will be deployed to the same environment,
```
Comment (eladb) [inline packages/@aws-cdk/aws-iam/lib/private/accounts.ts:19]: Mmmmmokay...

File: packages/@aws-cdk/aws-iam/lib/private/accounts.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+  // so treat as the same.
+  if (Token.isUnresolved(account1) && Token.isUnresolved(account2)) { return true; }
+
+  // One agnostic and the other one not means "shug".
```
Comment (eladb) [inline packages/@aws-cdk/aws-iam/lib/private/accounts.ts:23]:
```suggestion
  // One agnostic and the other one not means "shrug".
```

File: packages/@aws-cdk/aws-iam/lib/unknown-principal.ts [hunk 5]
```diff
@@ -43,4 +43,8 @@ export class UnknownPrincipal implements IPrincipal {
     this.resource.node.addWarning(`Add statement to this resource's role: ${repr}`);
     return true; // Pretend we did the work. The human will do it for us, eventually.
   }
+
+  public sameAccount(_scope: IConstruct): boolean | undefined {
```
Comment (eladb) [inline packages/@aws-cdk/aws-iam/lib/unknown-principal.ts:47]: Maybe return an enum instead of using `undefined` as a tri-state so this will be easier to reason about and avoid mistakes like (`!sameAccount())`).

File: packages/@aws-cdk/aws-iam/lib/principals.ts [hunk 6]
```diff
@@ -48,6 +48,17 @@ export interface IPrincipal extends IGrantable {
...
+  /**
+   * Return whether the principal and the given construct are in the same AWS account.
+   *
+   * This is used by `Grant` to determine whether a trust policy needs
+   * to be added to the resource as well as a permissions policy needs to be
+   * added to the identity.
+   *
+   * Returns `undefined` if it is unknown whether the accounts are the same.
+   */
+  sameAccount(scope: cdk.IConstruct): boolean | undefined;
```
Comment (eladb) [inline packages/@aws-cdk/aws-iam/lib/principals.ts:61]: 1. I feel that this API should accept a `cdk.Stack` and not a `cdk.IConstruct`. It will make this API much clearer.
2. Would it make sense to make this `_internal`? It's used only inside this module and TBH it feels like it's polluting the interface.

**Comment by skinny85** [comment]
> > I've actually changed my mind on this, and I want to explore an alternative solution. I think we should develop something here that's usable across all construct libraries, not specific to IAM.
> 
> I'm always in favor of better solutions for anything. As long as we can solve the same issue as mentioned here, go ahead.
> 
> > We have a related issue about this around the necessity of having access to `account` and `region` on a resource level, instead of just stack level, but I can't find it :/.
> 
> I guess because of the nature of our deployment mechanism:
> 
> * Environment of "owned" resources == environment of containing Stack
> * Environment of "imported" resources == ¯\_(ツ)_/¯ lol can be anything look at the ARN
> 
> That seems to be the crux of it, and you're right it holds for all imported resources, not just roles.
> 
> > @rix0rrr are you OK not merging this for the time being, and giving me some time to work on it?
> 
> As long as you incorporate the tests of this PR into whatever you come up with, sure!

@rix0rrr sorry this took so long, but this was a tricky one! I've opened #8280 with my proposed solution. It includes your tests from this PR.

I would appreciate if you could take a look, and let me know what you think!
